### PR TITLE
On narrowing, keep messages in streams and private messages in view.

### PR DESCRIPTION
**Pull request by Sam1301**
This pr addresses #145 
I have changed the way we are fetching messages when a message tile is clicked. Now, the message being clicked is used as an anchor while fetching and kept in view on narrowing.

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@Sam1301, it looks like you've already signed the Dropbox CLA. Thanks!

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java [hunk 1]
```diff
@@ -1645,6 +1659,12 @@ public void onNarrow(NarrowFilter narrowFilter) {
         doNarrow(narrowFilter);
     }
 
+
+    public void onNarrow(NarrowFilter narrowFilter, int messageId) {
+        // TODO: check if already narrowed to this particular stream/subject
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: What about this TODO?

**Comment by kunall17** [comment]
This one is a biggie but important as well, so what are your thoughts on this one @vishwesh3 @saketkumar95?
Can you guys give a quick review?

**Comment by vishwesh3** [comment]
sure

File: app/src/main/java/com/zulip/android/activities/ZulipActivity.java (hunk 1 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/ZulipActivity.java]: done, forgot to implement it :+1:

**Comment by Sam1301** [comment]
Thanks @vishwesh3, made the changes.

**Comment by Sam1301** [comment]
update: added support for narrowing to last message read in a particular stream when a stream is selected from the streams drawer.

**Comment by kunall17** [comment]
I tested the branch the but the commit f83d070 doesn't works, even if the messages are in the cache and read they aren't scrolled to the last read message! 
ezgif com-20435f322b

**Comment by Sam1301** [comment]
>Now, the message being clicked is used as an anchor while fetching and kept in view on narrowing.

I guess thats because this was a narrow on message click, but not from a stream in the stream drawer.

**Comment by kunall17** [comment]
Ok, didn't read the code before! 
This should be supported in these messageHeaderParent and narrow from the menu on long press on message! 

Works well from the streams drawer 👍

**Comment by Sam1301** [comment]
> This should be supported in these messageHeaderParent and narrow from the menu on long press on message!

sorry I didn't get that, can you please elaborate more on this? I have added code to support narrow to a message clicked keeping the message in view. This is done for messageHeaderParent. Are you suggesting to do the same for long press menu on message?

**Comment by kunall17** [comment]
These changes https://github.com/zulip/zulip-android/pull/279/files#diff-93e07d12c8f17453899c8a8cceb659d2L1157 should be here as well https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java#L130

**Comment by Sam1301** [comment]
hmm.. Is the reference correct ? I have changed https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java#L130 to https://github.com/zulip/zulip-android/pull/279/files#diff-6727cdd29f9d849bb181415059bd3279L130 to account for message in view on click.

**Comment by kunall17** [comment]
Nope this interface is implemented in ZulipActivity and this is the function https://github.com/zulip/zulip-android/blob/1229ae6afef8443f10d119edd6bf4ade1924d0f3/app/src/main/java/com/zulip/android/activities/ZulipActivity.java#L1798

**Comment by kunall17** [comment]
When I tried on my emulator and dev env it worked but when on zulip.tabbott.net it didn't load older messages! 

By the logs i saw 

> 01-31 22:42:48.777 23876-23876/com.zulip.android.dev I/onMessage: Already have 125414
> 01-31 22:42:48.777 23876-23876/com.zulip.android.dev I/onMessage: Already have 125415

Messages like these in large number! 

ezgif com-22e9fb136d

**Comment by kunall17** [comment]
Btw awesome work on finding the bug in 921fb86 👍

**Comment by Sam1301** [comment]
hmm.. its working an android 6.0.1 with chat.zulip.org. Can you please tell the android version on which you checked?

**Comment by kunall17** [comment]
It shouldn't matter with the android version, anyways its 6.0.1!

**Comment by Sam1301** [comment]
here's the [recording](https://drive.google.com/open?id=0B4jK5QX65b0ZanYxVXIyM2dYWDQ), working for me on android 6.0.1. Maybe some particular steps are involved to replicate this bug?

**Comment by kunall17** [comment]
Hmmm, checked again works now but I get many logs of already have still! 
Is that normal?

**Comment by kunall17** [comment]
@timabbott LGTM! :)

**Comment by timabbott** [comment]
Cool, merged, thanks @Sam1301 and @kunall17!  I did fix one typo in the first line of the last commit message (`poistion` to `position`).

**Pull request closed without merging**